mt-kahypar/partition/coarsening/multilevel_coarsener.h

```diff
@@ -371,60 +372,70 @@ class MultilevelCoarsener : public ICoarsener,
     // Will be important later for conflict resolution.
     bool success = false;
     const HypernodeWeight weight_u = hypergraph.nodeWeight(u);
-    HypernodeWeight weight_v = _cluster_weight[v];
+    HypernodeWeight weight_v = _cluster_weight[v].load(std::memory_order_relaxed);
     if ( weight_u + weight_v <= _context.coarsening.max_allowed_node_weight ) {
 
+      // We can use memory_order_relaxed when setting the state to MATCHING_IN_PROGRESS since this does not involve any writes to
+      // other locations, except for _matching_partner[u] (the latter is only relevant in conflict resolution and the resolution
+      // doesn't need stronger semantic, either). Setting the state to MATCHED however involves updating the cluster ID, for which
+      // we need acquire/release semantics
       uint8_t expect_unmatched_u = STATE(MatchingState::UNMATCHED);
-      if ( _matching_state[u].compare_exchange_strong(expect_unmatched_u, match_in_progress) ) {
-        _matching_partner[u] = v;
+      if ( _matching_state[u].compare_exchange_strong(expect_unmatched_u, match_in_progress, std::memory_order_relaxed) ) {
```

Comment:
In a compare-exchange, the memory ordering should be acquire

Reply:
Uhm, for what reason? This is absolutely not generally true. You can't just map operations to orderings, that is not how memory ordering semantics work

Reply:
Wouldn't some of the loads and stores in the locked state be allowed to moved before the CAS? Is that why the load on line 389 has acquire ordering? Does that prevent this?

Reply:
So, the accesses we need to be careful about are loads and stores of `cluster_ids`. New cluster IDs are written by the "owning" thread of the according node in the `joinCluster` operation. Cluster IDs are read either during conflict resolution (line 440) or when the state of `v` is already MATCHED (line 393). So in these two cases, we need to establish a happens-before relationship between the reading thread (lets say t_u) and the writing thread (t_v).

Line 393: Since u is owned by t_u, no other thread than t_u has performed a write on `matching_state[u]`. Thus, using acquire semantics for the `compare_exchange` wouldn't do anything here, since there is nothing to establish a happens-before relationship with. However, t_v has updated `matching_state[v]` to MATCHED after updating the cluster ID. What we need is therefore a happens-before relationship with regards to `matching_state[v]`. This is achieved by the release semantics in line 448 (as well as 403/430, though I think the release semantics aren't even strictly necessary there, since the cluster ID does not change) and the load on line 389 with acquire semantics.

Line 440: Same principle, in this case the acquire semantics on `matching_state[v]` are provided by the load in line 432.

Now, of course this does not provide any ordering guarantees for the accesses to `_matching_partner` (e.g. line 384). However, as discussed before, acquire semantics on the `compare_exchange` wouldn't help with this either, since there is nothing to synchronize against (I guess if we want that ordering we could use release semantics for the write in line 384). Luckily, the conflict resolution code does not need any ordering guarantees, since it just loops until a consistent state is detected. Only when the conflict is resolved and writes happen we need the acquire/release semantics.

So the summarized argument for the correctness is that (a) all accesses to `cluster_ids` are synchronized via acquire/release semantics on the read/write when  `cluster_ids[v]` changes state to MATCHED and (b) we don't need ordering guarantees for accesses to `_matching_partner`.

Of course there is also `_cluster_weight`, but here we just accept that a cluster might get too heavy if a race condition happens. To fix this we would need a `compare_exchange` loop in `joinCluster`, which we probably don't want.